# CPTAC3 GBM analysis
The repository contains all the additional preprocessing scripts/pipelines, and
analysis scripts to generate all the results and figures for CPTAC3
Giloblastoma (GBM) cohort.

For full information about this project, please refer to [the lab wiki][wiki]
for details.

[wiki]: https://confluence.ris.wustl.edu/pages/viewpage.action?pageId=37130883



## Discovery cohort
### `201901_locate_discovery_data` on denali
Locate the up-to-date GDC and proteomic data relaese of the discovery cohort
and the local file maps on denali, katmai, and MGI.


### `201901_gene_quantification` on katmai
Generate the transcript quantification (TPM) and gene-level expression of the
discovery cohort using Salmon.


### `201901_mutation_and_variant` on denali
Load the somatic mutations and germline variants of the discovery cohort
generated by WashU and GDC.


### `201902_somatic_wgs_cnv` on katmai
Collect the somatic CNV calls from WGS.


### `201903_case_review` on katmai
Review some cases that may not be glioblastoma.



## Proteomic QC
### `201811_locate_gdc_data` on denali
üìç *Use `201901_locate_discovery_data` instead*<br>
Locate the current GDC data release of CPTAC3 GBM cohort and their local
whereabouts.


### `201811_proteomic_qc` on denali
The preliminary analysis to help Tao from PNNL to design the proteomic
experiment plexes. The main goal is to address the 5 tumors that are clustered
with GTEx normals using single-shot mass spectrometry.


### `201811_rnaseq_salmon` on katmai
üìç *Use `201901_gene_quantification` instead*<br>
RNA quantification of high-confident protein coding trasncripts (Ensembl v90).
Only transcripts of protein coding with TSL 1 or TSL NA (e.g. single exon
trasncripts) were included. Since the mapping rate using this approach is low,
we also explore some other RNA-seq pipelines:

- Salmon using all transcripts
- Alignment-based pipeline: STAR + htseq


### `201811_wxs_bam_readcount` on katmai
WXS BAM readcount on regions of interest.
