---
title: "Annotate gene level CNV"
output: html_notebook
---

```{r, message=FALSE}
library(tidyverse)
library(ensembldb)
library(SummarizedExperiment)
```

```{r}
ENSDB_PTH <- '../../201901_gene_quantification/external_data/ensembl_ref/EnsDb.Hsapiens.v94.sqlite'
SEQINFO_PTH <- '../seqinfo_GRCh38.d1.vd1.rds'
```


## Read the gene level CNV of all samples
```{r}
sample_tbl <- read_tsv('../../201901_locate_discovery_data/tracked_results/CPTAC3_GBM_GDC_omics_UUIDs.tsv') %>%
    dplyr::filter_at(vars(wgs_blood_normal_BAM, wgs_tumor_BAM), all_vars(!is.na(.)))
```

```{r}
gene_cnv_tbl <- sample_tbl$case %>%
    purrr:::map(function(case) {
        pth = str_interp('../external_data/bicseq2_cnv/gene/${case}.tsv.gz')
        read_tsv(
            pth, 
            col_names = c('symbol', 'chrom', 'start', 'end', case),
            col_types = cols(
              symbol = col_character(),
              chrom = col_character(),
              .default = col_double()
            ) 
        )
    }) %>%
    purrr:::reduce(.f = ~ full_join(.x, .y, by = c('symbol', 'chrom', 'start', 'end')))
```

```{r}
gene_cnv_tbl %>% head()
```

## Read genome and annotation
SeqInfo object for GDC's genome 
```{r}
hg38_seqinfo <- readRDS(SEQINFO_PTH)
```

Read the Ensembl annotation and let it use UCSC chromosome naming style
```{r}
edb <- EnsDb(ENSDB_PTH)
seqlevelsStyle(edb) <- 'UCSC'
```


## Annotate the genes in the CNV call with the consistent Ensembl annotation.
We only select those genes on canonical chromosomes.
```{r}
edb_genes <- genes(
    edb, 
    filter = ~ gene_biotype != 'LRG_gene' & seq_name %in% str_c('chr', c(1:22, 'X', 'Y', 'M')),
    columns = c("seq_name", "gene_seq_start", "gene_seq_end", "seq_strand", "gene_id", "gene_id_version", "symbol"),
    return.type = "data.frame"
)
```

We query for the latest gene symbol and the Ensembl gene ID for each row.
```{r}
new_gene_annotation_tbl <- purrr::pmap_dfr(
    gene_cnv_tbl, 
    function(symbol, chrom, start, end, ...) {
        # Try to use the symbol name to find the gene
        g_start = start
        g_end = end
        g_symbol = symbol
        current_gene = edb_genes %>% subset(symbol == g_symbol)
        if (nrow(current_gene) == 1) {
            edb_chrom = current_gene$seq_name
            edb_start = current_gene$gene_seq_start
            edb_end = current_gene$gene_seq_end
            if ((edb_chrom == chrom) & (edb_start == g_start + 1) & (edb_end == g_end)) {
                return(current_gene)
            } 
        }
        # Use location to find the gene
        current_gene = edb_genes %>% subset(seq_name == chrom & gene_seq_start == g_start + 1 & gene_seq_end == g_end)
        if (nrow(current_gene) > 1) {
            cat(symbol, 'has more than one entry in EnsDb\n')
            return(NA) 
        } else if (nrow(current_gene) == 0) {
            cat(symbol, 'not found\n')
            return(NA)
        }
        return(current_gene)
    }
) %>% as_tibble()
```

Add back the original symbol name
```{r}
row_annotation_tbl <- new_gene_annotation_tbl %>%
    mutate(
        seq_strand = case_when(
            seq_strand == -1 ~ '-',
            seq_strand == 1 ~ '+',
            TRUE ~ '*'
        ),
        original_symbol = gene_cnv_tbl$symbol
    ) %>%
    dplyr::select(-gene_biotype)
```

The symbol names are identical
```{r}
row_annotation_tbl %>%
    dplyr::filter(symbol != original_symbol)
```


```{r}
row_gr <- makeGRangesFromDataFrame(
    row_annotation_tbl %>% column_to_rownames(var = 'symbol'),
    seqnames.field = 'seq_name',
    start.field = 'gene_seq_start',
    end.field = 'gene_seq_end',
    strand.field = 'seq_strand',
    keep.extra.columns = TRUE,
    seqinfo = hg38_seqinfo
)
row_gr
```

```{r}
gene_cnv_mat <- as.matrix(gene_cnv_tbl[, -(1:4)])
gene_cnv_se <- SummarizedExperiment(
    rowRanges = row_gr,
    assays = list(
        cnv_log2 = gene_cnv_mat
    ),
    metadata = list(
        cohort = 'CPTAC3 GBM discovery cohort (incomplete)',
        description = 'Gene level somatic CNV (log2) based on whole genome sequencing',
        pipeline = 'BIC-seq2 https://github.com/ding-lab/BICSEQ2',
        annotation = 'GENCODE v29 (Ensembl v94) protein coding only',
        contact = 'Liang-Bo Wang <liang-bo.wang@wustl.edu>'
    )
)
```

```{r}
saveRDS(gene_cnv_se, '../processed_data/gene_cnv.rds')
```



