---
title: "Download raw sequencing data"
output: html_notebook
---

```{r, message=FALSE}
library(tidyverse)
```

```{r}
cohort_seq_tbl <- read_tsv('../tracked_results/cptac3_gbm_seq_status.tsv')
```


```{r}
gbm_seq_data <- read_tsv(
    '../../matt_catalog/CPTAC3.AR.dat',
    col_types = cols(.default = col_character(), filesize = col_double())
) %>%
    rename(sample_name = `# sample_name`) %>%
    filter(disease == 'GBM') 
```

```{r}
bam_maps <- list.files('../../matt_catalog', '.*.BamMap.dat$', full.names = TRUE) %>%
    set_names(str_match(., '.*/([^.]+).BamMap.dat$')[, 2]) %>%
    map(function(pth) {
        read_tsv(pth, col_types = cols(.default = col_character(), filesize = col_double())) %>%
            rename(sample_name = `# sample_name`) %>%
            filter(disease == 'GBM')
    })
```



## Download the RNA-seq to MGI
```{r}
cases_with_rna_processed <- read_lines('../genomic_downstream_availability/cases_with_rna_processed.list') %>%
    setdiff('C3N-01850')  # the sample is being re-sequenced; don't use the current one
```

```{r}
rna_to_download <- gbm_seq_data %>%
    filter(experimental_strategy == 'RNA-Seq', data_format == 'FASTQ',
           !case %in% cases_with_rna_processed)
```

```{r}
cohort_seq_tbl %>%
    filter(case %in% rna_to_download$case)
```


## Download WGS to MGI
```{r}
cases_in_harm_batch3 = read_lines('../tracked_results/wxs_or_wgs_unharmonized_cases.list')
```

```{r}
cases_with_gdc_bam_cnv = read_lines('../genomic_downstream_availability/cases_with_gdc_harmonized_wgs_somatic_cnv.list')
```

Make sure all tumors in the discvoery cohort are covered:
```{r}
cohort_seq_tbl %>% 
    filter(in_discovery, !is_gtex_normal) %>%
    pull(case) %>%
    setdiff(cases_with_gdc_bam_cnv) %>% 
    setdiff(cases_in_harm_batch3)
```


```{r}
wgs_to_download <- gbm_seq_data %>%
    filter(experimental_strategy == 'WGS', data_format == 'BAM', reference == 'hg38', 
           case %in% cases_in_harm_batch3)

# all cases have paired tumor and normal WGS BAMs
nrow(wgs_to_download) == length(cases_in_harm_batch3) * 2
```


## Download WXS to MGI
```{r}
cases_with_gdc_wxs <- cohort_seq_tbl %>% filter(has_wxs_normal_gdc) %>% 
    pull(case) %>% unique()

washu_wxs_bam_map <- read_tsv(
    '../../201904_locate_adhoc_data/tracked_results/MGI.GBM_custom_wxs.BamMap.dat',
     col_types = cols(.default = col_character(), filesize = col_double())
)
```

```{r}
wxs_to_download <- gbm_seq_data %>%
    filter(experimental_strategy == 'WXS', data_format == 'BAM', reference == 'hg19', 
           !case %in% cases_with_gdc_wxs) %>%
    anti_join(washu_wxs_bam_map, by = c('case', 'sample_type'))
```


```{r}
files_to_download <- bind_rows(
    # rna_to_download,
    wgs_to_download,
    # wxs_to_download
)
files_to_download %>% 
    group_by(experimental_strategy) %>%
    summarize(filesize_GB = round(sum(filesize) / (2 ** 30), 1))
```

```{r}
files_to_download %>%
    write_tsv('~/Downloads/gdc_seq_files_to_download.tsv.gz')
```

