---
title: "Check the availability of genomic data on the lab servers"
output: html_notebook
---

```{r}
library(tidyverse)
```


## Discovery samples
```{r}
discovery_samples <- read_lines('../cptac3_gbm_discovery_samples.list')
discovery_samples_cptac_only <- str_subset(discovery_samples, '^C3[NL]-')
discovery_samples_gtex_only <- str_subset(discovery_samples, '^PT-')
```

Read the GDC UUIDs 
```{r}
gdc_omics_uuid_tbl <- read_tsv('../tracked_results/CPTAC3_GBM_GDC_omics_UUIDs.tsv')
```

```{r}
local_data_location_tbl <- map_dfr(
    c('../../matt_catalog/MGI.BamMap.dat', 
      '../../matt_catalog/katmai.BamMap.dat'),
    function(x) {
        read_tsv(
            x, 
            progress = FALSE,
            col_types =  cols(
                `# sample_name` = col_character(),
                case = col_character(),
                disease = col_character(),
                experimental_strategy = col_character(),
                sample_type = col_character(),
                data_path = col_character(),
                filesize = col_double(),
                data_format = col_character(),
                reference = col_character(),
                UUID = col_character(),
                system = col_character()
            )
        ) %>%
            rename(data_name = `# sample_name`) %>%
            filter(case %in% discovery_samples_cptac_only)
    }
) 
```


### hg38 WXS BAMs on MGI
```{r}
mgi_wxs_bams <- local_data_location_tbl %>% filter(system == 'MGI' & experimental_strategy == 'WXS') %>%
    select(data_name, case, experimental_strategy, sample_type, UUID, data_format, system, data_path)

nrow(mgi_wxs_bams)
```


### hg38 WGS BAMs on katmai
```{r}
katmai_wgs_bams <- local_data_location_tbl %>% filter(system == 'katmai' & experimental_strategy == 'WGS') %>%
    select(data_name, case, experimental_strategy, sample_type, UUID, data_format, system, data_path)

nrow(katmai_wgs_bams)
```

`C3L-01834` only have normal WGS BAM:
```{r}
katmai_wgs_bams %>% 
    select(case, sample_type, UUID) %>%
    spread(sample_type, UUID) %>%
    filter_at(vars(c('blood_normal', 'tumor')), any_vars(is.na(.)))
```


### RNA-seq FASTQs on katmai
```{r}
katmai_rna_fqs <- local_data_location_tbl %>% 
    filter(system == 'katmai' & experimental_strategy == 'RNA-Seq') %>%
    select(data_name, case, experimental_strategy, sample_type, UUID, data_format, system, data_path)

katmai_rna_fqs %>% nrow()
```


### Compare with the previous 60 samples
```{r}
samples_already_on_gdc <- read_tsv(
    '../../matt_catalog/CPTAC3.cases.dat',
    col_names = c('case_id', 'cancer_type')
) %>% 
    filter(cancer_type == 'GBM', case_id %in% discovery_samples_cptac_only) %>% 
    pull(case_id) %>% 
    sort()

prev_60_genomic_samples <- read_lines('../../201811_locate_gdc_data/genomic_samples.60_cases.list')
```

List all the samples that we can download:
```{r}
setdiff(samples_already_on_gdc, prev_60_genomic_samples)
```

